# Steps Involved in Data Cleaning & Preprocessing:
    1) Importing Libraries and Reading the Dataframes
    2) Merging the DataFrames
    3) Removing the Duplicates
    4) Checking for Null values and types of data
    5) Replacing suitable values for null values
    6) Changing the Datatype
    7) Creating and formatting necessary Coulumns
    8) Normalising the data
    9) Quality Check

## Importing Libraries and Reading the Dataframes

In [3]:
# Importing librariee
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [4]:
# Reading csv and creatig data frame
# April may DataFrame as dfam
dfam = pd.read_csv('D:/UpGrad/Cult Fit Project/Fitness_Classes_Data/Fitness Classes Data/Classes April-May 2018.csv')
dfam.head()

ActivitySiteID      ActivityDescription  \
0            HXP  20-20-20  2.45pm-3.45pm   
1            HXP  20-20-20  2.45pm-3.45pm   
2            HXP  20-20-20  2.45pm-3.45pm   
3            HXP  20-20-20  2.45pm-3.45pm   
4            HXP  20-20-20  2.45pm-3.45pm   

  BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
0                               08-Apr-18         14:45:00          25   
1                               15-Apr-18         14:45:00          25   
2                               22-Apr-18         14:45:00          25   
3                               29-Apr-18         14:45:00          25   
4                               06-May-18         14:45:00          25   

   Number Booked  Price (INR)  
0             12        499.0  
1             15        499.0  
2             14        499.0  
3              9        499.0  
4              7        499.0

In [5]:
dfam.shape

(2177, 7)

April and May dataframe has 7 columns and 2177 rows

In [7]:
# June DataFrame as dfj
dfj = pd.read_csv('D:/UpGrad/Cult Fit Project/Fitness_Classes_Data/Fitness Classes Data/Classes June 2018.csv')
dfj.head()

ActivitySiteID     ActivityDescription  \
0            BRP  20:20:20  9.30-10.30am   
1            BRP  20:20:20  9.30-10.30am   
2            BRP  20:20:20  9.30-10.30am   
3            BRP  20:20:20  9.30-10.30am   
4            BRP  20:20:20  9.30-10.30am   

  BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
0                               01-Jun-18          9:30:00          35   
1                               08-Jun-18          9:30:00          35   
2                               15-Jun-18          9:30:00          35   
3                               22-Jun-18          9:30:00          35   
4                               29-Jun-18          9:30:00          35   

   Number Booked  Price (INR)  
0             28          499  
1             35          499  
2             31          499  
3             32          499  
4             19          499

In [8]:
dfj.shape

(1112, 7)

June dataframe has 7 columns and 1112 rows

## Merging the DataFrames

In [11]:
# df as merged dataframe
df = pd.concat([dfam,dfj],axis=0)
df.tail()

ActivitySiteID ActivityDescription  \
1107            TSC  Zumba  6.30-7.30pm   
1108            TSC  Zumba  6.30-7.30pm   
1109            TSC       Zumba - 6-7pm   
1110            TSC       Zumba - 6-7pm   
1111            TSC       Zumba - 6-7pm   

     BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
1107                               18-Jun-18         18:30:00          18   
1108                               25-Jun-18         18:30:00          18   
1109                               07-Jun-18         18:00:00          50   
1110                               14-Jun-18         18:00:00          50   
1111                               21-Jun-18         18:00:00          50   

      Number Booked  Price (INR)  
1107              9       1299.0  
1108              9       1299.0  
1109             16       1299.0  
1110             13       1299.0  
1111             12       1299.0

In [12]:
df.head()

ActivitySiteID      ActivityDescription  \
0            HXP  20-20-20  2.45pm-3.45pm   
1            HXP  20-20-20  2.45pm-3.45pm   
2            HXP  20-20-20  2.45pm-3.45pm   
3            HXP  20-20-20  2.45pm-3.45pm   
4            HXP  20-20-20  2.45pm-3.45pm   

  BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
0                               08-Apr-18         14:45:00          25   
1                               15-Apr-18         14:45:00          25   
2                               22-Apr-18         14:45:00          25   
3                               29-Apr-18         14:45:00          25   
4                               06-May-18         14:45:00          25   

   Number Booked  Price (INR)  
0             12        499.0  
1             15        499.0  
2             14        499.0  
3              9        499.0  
4              7        499.0

In [13]:
df.shape

(3289, 7)

merged dataframe has 7 columns and 3289 rows

## Removing the Duplicates

In [16]:
df = df.drop_duplicates()
df.shape

(3289, 7)

No duplicates found in the Dataframe

## Checking for Null values and types of data

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3289 entries, 0 to 1111
Data columns (total 7 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   ActivitySiteID                           3289 non-null   object 
 1   ActivityDescription                      3289 non-null   object 
 2   BookingEndDateTime (Month / Day / Year)  3289 non-null   object 
 3   BookingStartTime                         3289 non-null   object 
 4   MaxBookees                               3289 non-null   int64  
 5   Number Booked                            3289 non-null   int64  
 6   Price (INR)                              3271 non-null   float64
dtypes: float64(1), int64(2), object(4)
memory usage: 205.6+ KB


In [20]:
df.isnull().sum()

ActivitySiteID                              0
ActivityDescription                         0
BookingEndDateTime (Month / Day / Year)     0
BookingStartTime                            0
MaxBookees                                  0
Number Booked                               0
Price (INR)                                18
dtype: int64

In [21]:
df.describe()

MaxBookees  Number Booked  Price (INR)
count  3289.000000    3289.000000  3271.000000
mean     32.179386      17.088781  1852.714460
std      15.092156       9.504098   792.636673
min       2.000000       1.000000   499.000000
25%      24.000000      10.000000  1299.000000
50%      30.000000      16.000000  1499.000000
75%      35.000000      24.000000  2499.000000
max      70.000000      64.000000  3999.000000

Null Values not found in the data frame except for the price column
price column has 18 null values present.

Next we will understand how we can compute the missing values, removing the null values can be kept as lst option

In [23]:
df[df['Price (INR)'].isnull()]

ActivitySiteID      ActivityDescription  \
84             TSC    5.30-6.30pm Body Cond   
85             TSC    5.30-6.30pm Body Cond   
86             TSC    5.30-6.30pm Body Cond   
87             TSC    5.30-6.30pm Body Cond   
88             TSC    5.30-6.30pm Body Cond   
89             TSC    5.30-6.30pm Body Cond   
90             TSC    5.30-6.30pm Body Cond   
91             TSC    5.30-6.30pm Body Cond   
92             TSC    5.30-6.30pm Body Cond   
93             TSC  6.30-7.30pm Combat Aero   
94             TSC  6.30-7.30pm Combat Aero   
95             TSC  6.30-7.30pm Combat Aero   
96             TSC  6.30-7.30pm Combat Aero   
97             TSC  6.30-7.30pm Combat Aero   
98             TSC  6.30-7.30pm Combat Aero   
99             TSC  6.30-7.30pm Combat Aero   
100            TSC  6.30-7.30pm Combat Aero   
101            TSC  6.30-7.30pm Combat Aero   

    BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
84                                03-Apr-18         17:30:00          24   
85                                10-Apr-18         17:30:00          24   
86                                17-Apr-18         17:30:00          24   
87                                24-Apr-18         17:30:00          24   
88                                01-May-18         17:30:00          24   
89                                08-May-18         17:30:00          24   
90                                15-May-18         17:30:00          24   
91                                22-May-18         17:30:00          24   
92                                29-May-18         17:30:00          24   
93                                03-Apr-18         18:30:00          24   
94                                10-Apr-18         18:30:00          24   
95                                17-Apr-18         18:30:00          24   
96                                24-Apr-18         18:30:00          24   
97                                01-May-18         18:30:00          24   
98                                08-May-18         18:30:00          24   
99                                15-May-18         18:30:00          24   
100                               22-May-18         18:30:00          24   
101                               29-May-18         18:30:00          24   

     Number Booked  Price (INR)  
84               7          NaN  
85              14          NaN  
86              12          NaN  
87              11          NaN  
88              13          NaN  
89               9          NaN  
90               5          NaN  
91              13          NaN  
92               9          NaN  
93               9          NaN  
94               8          NaN  
95              11          NaN  
96              14          NaN  
97              12          NaN  
98              10          NaN  
99               9          NaN  
100             10          NaN  
101             10          NaN

In [24]:
df[df['ActivityDescription'] == '5.30-6.30pm Body Cond']

ActivitySiteID    ActivityDescription  \
84             TSC  5.30-6.30pm Body Cond   
85             TSC  5.30-6.30pm Body Cond   
86             TSC  5.30-6.30pm Body Cond   
87             TSC  5.30-6.30pm Body Cond   
88             TSC  5.30-6.30pm Body Cond   
89             TSC  5.30-6.30pm Body Cond   
90             TSC  5.30-6.30pm Body Cond   
91             TSC  5.30-6.30pm Body Cond   
92             TSC  5.30-6.30pm Body Cond   
936            TSC  5.30-6.30pm Body Cond   
937            TSC  5.30-6.30pm Body Cond   
938            TSC  5.30-6.30pm Body Cond   
939            TSC  5.30-6.30pm Body Cond   

    BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
84                                03-Apr-18         17:30:00          24   
85                                10-Apr-18         17:30:00          24   
86                                17-Apr-18         17:30:00          24   
87                                24-Apr-18         17:30:00          24   
88                                01-May-18         17:30:00          24   
89                                08-May-18         17:30:00          24   
90                                15-May-18         17:30:00          24   
91                                22-May-18         17:30:00          24   
92                                29-May-18         17:30:00          24   
936                               05-Jun-18         17:30:00          24   
937                               12-Jun-18         17:30:00          24   
938                               19-Jun-18         17:30:00          24   
939                               26-Jun-18         17:30:00          24   

     Number Booked  Price (INR)  
84               7          NaN  
85              14          NaN  
86              12          NaN  
87              11          NaN  
88              13          NaN  
89               9          NaN  
90               5          NaN  
91              13          NaN  
92               9          NaN  
936             14       1299.0  
937             13       1299.0  
938              7       1299.0  
939              2       1299.0

In [25]:
df[df['ActivityDescription'] == '6.30-7.30pm Combat Aero']

ActivitySiteID      ActivityDescription  \
93             TSC  6.30-7.30pm Combat Aero   
94             TSC  6.30-7.30pm Combat Aero   
95             TSC  6.30-7.30pm Combat Aero   
96             TSC  6.30-7.30pm Combat Aero   
97             TSC  6.30-7.30pm Combat Aero   
98             TSC  6.30-7.30pm Combat Aero   
99             TSC  6.30-7.30pm Combat Aero   
100            TSC  6.30-7.30pm Combat Aero   
101            TSC  6.30-7.30pm Combat Aero   
940            TSC  6.30-7.30pm Combat Aero   
941            TSC  6.30-7.30pm Combat Aero   
942            TSC  6.30-7.30pm Combat Aero   
943            TSC  6.30-7.30pm Combat Aero   

    BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
93                                03-Apr-18         18:30:00          24   
94                                10-Apr-18         18:30:00          24   
95                                17-Apr-18         18:30:00          24   
96                                24-Apr-18         18:30:00          24   
97                                01-May-18         18:30:00          24   
98                                08-May-18         18:30:00          24   
99                                15-May-18         18:30:00          24   
100                               22-May-18         18:30:00          24   
101                               29-May-18         18:30:00          24   
940                               05-Jun-18         18:30:00          24   
941                               12-Jun-18         18:30:00          24   
942                               19-Jun-18         18:30:00          24   
943                               26-Jun-18         18:30:00          24   

     Number Booked  Price (INR)  
93               9          NaN  
94               8          NaN  
95              11          NaN  
96              14          NaN  
97              12          NaN  
98              10          NaN  
99               9          NaN  
100             10          NaN  
101             10          NaN  
940              6       1499.0  
941              6       1499.0  
942             10       1499.0  
943              5       1499.0

Prices are missing in April and may data set but are present in the June Dataset.

Prices for each Activities with deifferent location and time period are same

Price for Bondy Cont in june is 1299 and we can imput these value for missing Body Cond, since people have booked classes for te same.

Price for Bondy Cont in june is 1499 and we can imput these value for missing Combat Aero, since people have booked classes for te same.

## Replacing suitable values for null values

In [28]:
df.loc[df['ActivityDescription'] == '5.30-6.30pm Body Cond','Price (INR)'] = 1299 # replacing null with the median price of Body Cond

In [29]:
df.loc[df['ActivityDescription'] == '6.30-7.30pm Combat Aero','Price (INR)'] = 1499 # ewplacing null with Combat Aero

In [30]:
df.isnull().sum() # confirming the rectified dataframe

ActivitySiteID                             0
ActivityDescription                        0
BookingEndDateTime (Month / Day / Year)    0
BookingStartTime                           0
MaxBookees                                 0
Number Booked                              0
Price (INR)                                0
dtype: int64

## Changing the Datatype

In [32]:
df['BookingEndDateTime (Month / Day / Year)'] = pd.to_datetime(df['BookingEndDateTime (Month / Day / Year)'],format='%d-%b-%y')

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3289 entries, 0 to 1111
Data columns (total 7 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   ActivitySiteID                           3289 non-null   object        
 1   ActivityDescription                      3289 non-null   object        
 2   BookingEndDateTime (Month / Day / Year)  3289 non-null   datetime64[ns]
 3   BookingStartTime                         3289 non-null   object        
 4   MaxBookees                               3289 non-null   int64         
 5   Number Booked                            3289 non-null   int64         
 6   Price (INR)                              3289 non-null   float64       
dtypes: datetime64[ns](1), float64(1), int64(2), object(3)
memory usage: 205.6+ KB


## Creating and formatting necessary Coulumns

In [35]:
df.shape

(3289, 7)

In [36]:
# formatting Activity Description value in common format
df['ActivityDescription'] = df['ActivityDescription'].astype(str)

mask_main = df['ActivityDescription'].str.match(r'^\d')

mask_exc = (
    df['ActivityDescription'].str.startswith('20-') |
    df['ActivityDescription'].str.startswith('20:') |
    df['ActivityDescription'].str.startswith('45')
)

df_result = df[mask_main & ~mask_exc]


In [37]:
# Values which have abnormal format
df_result['ActivityDescription'].unique()

array(['5.30-6.30pm Body Cond', '6.30-7.30pm Combat Aero'], dtype=object)

In [38]:
nameswap =  {'5.30-6.30pm Body Cond':'Body Cond 5.30-6.30pm','6.30-7.30pm Combat Aero':'Combat Aero 6.30-7.30pm'} # new format dict

In [39]:
# Renaming
df['ActivityDescription'] = df['ActivityDescription'].replace(nameswap)

Creating a cleaned New Column called Activity


In [41]:
# Splitting the Activity Description Columns
ndf = df.copy()
ndf['new'] = ndf['ActivityDescription'].apply(lambda x: x.split(' ')[0])
ndf['new2'] = ndf['ActivityDescription'].apply(lambda x: x.split(' ')[1] if len(x.split(' ')) >= 2 else ' ' )
ndf['new2'] = ndf['new2'].astype(str)
ndf['new3'] = ndf['ActivityDescription'].apply(lambda x: x.split(' ')[2] if len(x.split(' ')) >= 3 else ' ' )
ndf['new3'] = ndf['new3'].astype(str)
ndf['new2'] = ndf['new2'].apply(
    lambda x: ' ' if x.startswith(('1','2','3','4','5','6','7','8','9','10','11','12')) else x)
ndf['new3'] = ndf['new3'].apply(
    lambda x: ' ' if x.startswith(('1','2','3','4','5','6','7','8','9','10','11','12')) or x.startswith('-') else x)
ndf['new4'] = ndf['ActivityDescription'].apply(lambda x: x.split(' ')[3] if len(x.split(' ')) >= 4 else ' ' )
ndf['new4'] = ndf['new4'].astype(str)
ndf['new4'] = ndf['new4'].apply(
    lambda x: ' ' if x.startswith(('1','2','3','4','5','6','7','8','9','10','11','12')) or x.startswith('-') else x)
ndf['new5'] = ndf['ActivityDescription'].apply(lambda x: x.split(' ')[4] if len(x.split(' ')) >= 5 else ' ' )
ndf['new5'] = ndf['new5'].astype(str)
ndf['new5'] = ndf['new5'].apply(
    lambda x: ' ' if x.startswith(('1','2','3','4','5','6','7','8','9','10','11','12')) or x.startswith('-') else x)

In [42]:
# joining the Splitted Columns and Stripping spaces
col = [x for x in ndf.columns if x.startswith('new')]
ndf = ndf[col]
ndf['Activity_uc'] = ndf['new'] + ndf['new2'] + ndf['new3'] + ndf['new4'] + ndf['new5']
ndf['Activity_uc']= ndf['Activity_uc'].str.strip()
ndf['Activity_uc'] = ndf['Activity_uc'].str.lower()
ndf['Activity_uc'] = ndf['Activity_uc'].astype(str)

In [43]:
# Renaming and Standardising Columns 
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: '20-20-20' if x.startswith('20') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Body Conditioning' if x.startswith('bodycon') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Aqua Babies' if x.startswith('aquaba') or x.startswith('aquat')  else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Aqua Fit' if x.startswith('aquafi')  else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Combat Aero' if x.startswith('combataer')  else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Barre Pilates' if x.startswith('barpila') or x.startswith('barrepil') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Better Living' if x.startswith('betterlivin') or x.startswith('exbetterlivin') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Body Attack' if x.startswith('bodyatta') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Body Balance' if x.startswith('bodybal') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Circuit' if x.startswith('circuit') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Junior Gym' if x.startswith('jnrgym') or x.startswith('juniorgym') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Pilates' if x.startswith('pilat') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Body Combat' if x.startswith('bodycomb') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Stretching' if x.startswith('stretc') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Fitness Pilates' if x.startswith('fitnespila') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Stretching' if x.startswith('basicstretchclass') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Zumba' if x.startswith('zumb') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Leg Bumtum' if x.startswith('leg') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'h.i.i.t' if x.startswith('h.i') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Hattonabc' if x.startswith('hatton') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Tai Chi' if x.startswith('tai') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Aqua Fit' if x.startswith('swim') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Stoke Skill' if x.startswith('stroke') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Nice and Easy' if x.startswith('nice') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Blt Activities' if x.startswith('blt') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Studio Cycling' if x.startswith('studioc') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Line Dancing' if x.startswith('lineda') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Step' if x.startswith('step') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: '45 Min Madness' if x.startswith('45') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'ABS Blast' if x.startswith('abs') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Aero' if x.startswith('aero') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Aqua Zumba' if x.startswith('aquazu') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Body Sculp Aero' if x.startswith('bodscu') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Ball Room Dance' if x.startswith('ballroom') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Body Pump' if x.startswith('bodyp') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Chair Based Exercise' if x.startswith('chairb') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Fusion' if x.startswith('fusion') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Gentle Keep Fit' if x.startswith('gentlekee') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Hiit Circuit' if x.startswith('hiitc') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Insanity' if x.startswith('insani') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Junior Induction' if x.startswith('junindu') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Junior Swimfit' if x.startswith('juniorswi') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'MMA Cond' if x.startswith('mma') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Meta Fit' if x.startswith('meta') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Power Yoga' if x.startswith('poweryo') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Body pump' if x.startswith('pump') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Stretching' if x.startswith('purestr') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Rock Roll Dancing' if x.startswith('rockrol') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Shabam' if x.startswith('sh') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Stretching' if x.startswith('strength&s') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Tap Dancing' if x.startswith('tapdan') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Yoga' if x.startswith('yoga') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Family BootCamp' if x.startswith('family') else x)
ndf['Activity_uc'] = ndf['Activity_uc'].apply(lambda x: 'Better Living' if x.startswith('wmbetter') else x)

In [44]:
# Merging the Column
df['Activity']= ndf['Activity_uc']
df.head()

ActivitySiteID      ActivityDescription  \
0            HXP  20-20-20  2.45pm-3.45pm   
1            HXP  20-20-20  2.45pm-3.45pm   
2            HXP  20-20-20  2.45pm-3.45pm   
3            HXP  20-20-20  2.45pm-3.45pm   
4            HXP  20-20-20  2.45pm-3.45pm   

  BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
0                              2018-04-08         14:45:00          25   
1                              2018-04-15         14:45:00          25   
2                              2018-04-22         14:45:00          25   
3                              2018-04-29         14:45:00          25   
4                              2018-05-06         14:45:00          25   

   Number Booked  Price (INR)  Activity  
0             12        499.0  20-20-20  
1             15        499.0  20-20-20  
2             14        499.0  20-20-20  
3              9        499.0  20-20-20  
4              7        499.0  20-20-20

In [45]:
# Creating a new Column 'Day'
df['Day'] = df['BookingEndDateTime (Month / Day / Year)'].dt.day_name()

In [46]:
# Creating a new Column 'Week'
df['Week'] = df['BookingEndDateTime (Month / Day / Year)'].dt.isocalendar().week

In [47]:
# Creating a new Column 'Month'
df['Month'] = df['BookingEndDateTime (Month / Day / Year)'].dt.month

In [48]:
# Checking MaxBookees Coulmn
df['MaxBookees'].value_counts()

MaxBookees
35    985
60    402
25    356
24    322
27    319
10    265
30    114
20     90
70     86
15     85
40     70
50     43
2      37
21     26
18     23
4      18
12     15
14     13
29     13
6       7
Name: count, dtype: int64

In [49]:
df[(df['MaxBookees']<10) & (df['Number Booked']>10)] # checkeing data frame where MaxBookees are leess but Number Booked is more

ActivitySiteID   ActivityDescription  \
265            SBP  Aquafit  7.00-8.00pm   
266            SBP  Aquafit  7.00-8.00pm   
267            SBP  Aquafit  7.00-8.00pm   
268            SBP  Aquafit  7.00-8.00pm   
269            SBP  Aquafit  7.00-8.00pm   
270            SBP  Aquafit  7.00-8.00pm   
271            SBP  Aquafit  7.00-8.00pm   

    BookingEndDateTime (Month / Day / Year) BookingStartTime  MaxBookees  \
265                              2018-04-06         19:00:00           6   
266                              2018-04-13         19:00:00           6   
267                              2018-04-20         19:00:00           6   
268                              2018-04-27         19:00:00           6   
269                              2018-05-04         19:00:00           6   
270                              2018-05-11         19:00:00           6   
271                              2018-05-18         19:00:00           6   

     Number Booked  Price (INR)  Activity     Day  Week  Month  
265             19        999.0  Aqua Fit  Friday    14      4  
266             25        999.0  Aqua Fit  Friday    15      4  
267             26        999.0  Aqua Fit  Friday    16      4  
268             31        999.0  Aqua Fit  Friday    17      4  
269             27        999.0  Aqua Fit  Friday    18      5  
270             29        999.0  Aqua Fit  Friday    19      5  
271             26        999.0  Aqua Fit  Friday    20      5

In [50]:
# Creating a Column 'Occupancy'
df['Occupancy'] = df['Number Booked']/df['MaxBookees']
df['Occupancy'] = df['Occupancy'].round(2)

In [51]:
# Creating a Column 'Revenue'
df['Revenue'] = df['Price (INR)']*df['Number Booked']

## Normalising the Data

In [53]:
from sklearn.preprocessing import StandardScaler
norm_df = df.copy()
num_cols = ['MaxBookees','Number Booked','Price (INR)','Occupancy','Revenue']
scaler = StandardScaler()
norm_df[num_cols] = scaler.fit_transform(norm_df[num_cols])
norm_df[num_cols]

MaxBookees  Number Booked  Price (INR)  Occupancy   Revenue
0      -0.475775      -0.535512    -1.708069  -0.287162 -1.190467
1      -0.475775      -0.219810    -1.708069   0.089699 -1.116867
2      -0.475775      -0.325044    -1.708069  -0.035922 -1.141400
3      -0.475775      -0.851213    -1.708069  -0.664024 -1.264066
4      -0.475775      -1.061680    -1.708069  -0.915264 -1.313133
...          ...            ...          ...        ...       ...
1107   -0.939663      -0.851213    -0.696802  -0.224352 -0.910080
1108   -0.939663      -0.851213    -0.696802  -0.224352 -0.910080
1109    1.180966      -0.114577    -0.696802  -0.789644 -0.463025
1110    1.180966      -0.430278    -0.696802  -0.978075 -0.654620
1111    1.180966      -0.535512    -0.696802  -1.040885 -0.718485

[3289 rows x 5 columns]

## Quality Check

In [55]:
def Quality_check(df):
    d =dict(df.isnull().sum())
    for x in d:
        if d[x] == 0:
            print(f'No null value Found in {x}')
            print(f'datatype {df[x].dtype}')
        else:
            print(f'Null value found in {x}')
            print(f'datatype {df[x].dtype}')

In [56]:
Quality_check(df)

No null value Found in ActivitySiteID
datatype object
No null value Found in ActivityDescription
datatype object
No null value Found in BookingEndDateTime (Month / Day / Year)
datatype datetime64[ns]
No null value Found in BookingStartTime
datatype object
No null value Found in MaxBookees
datatype int64
No null value Found in Number Booked
datatype int64
No null value Found in Price (INR)
datatype float64
No null value Found in Activity
datatype object
No null value Found in Day
datatype object
No null value Found in Week
datatype UInt32
No null value Found in Month
datatype int32
No null value Found in Occupancy
datatype float64
No null value Found in Revenue
datatype float64


In [57]:
# Creating a Excel file
with pd.ExcelWriter('D:/UpGrad/Cult Fit Project/Solution/Cleaned_classes.xlsx') as writer:
    df.to_excel(writer, sheet_name='Cleaned data', index=False)
    norm_df.to_excel(writer, sheet_name='Normalized data', index=False)

In [58]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3289 entries, 0 to 1111
Data columns (total 13 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   ActivitySiteID                           3289 non-null   object        
 1   ActivityDescription                      3289 non-null   object        
 2   BookingEndDateTime (Month / Day / Year)  3289 non-null   datetime64[ns]
 3   BookingStartTime                         3289 non-null   object        
 4   MaxBookees                               3289 non-null   int64         
 5   Number Booked                            3289 non-null   int64         
 6   Price (INR)                              3289 non-null   float64       
 7   Activity                                 3289 non-null   object        
 8   Day                                      3289 non-null   object        
 9   Week                                     3289 